### DISCUSSION TITLE
👝 Wallets

### DISCUSSION THREAD

**Discussion by shanejearley** [consensusnetworks/casimir#100] [General]
Notes on wallets

**Comment by shanejearley** [comment]
@ccali11 Something I stumbled on through other node work – MetaMask currently doesn't support **programmatic** switching of the RPC URL for default chains (mainnet, goerli, etc.). You can see developer requests to remove their restriction [here](https://github.com/MetaMask/metamask-extension/blob/develop/app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js#L143). We are interested because we'll have a use-case to switch the mainnet RPC URL to Casimir nodes, say if we want to pursue block building from our application. [Here](https://github.com/MetaMask/metamask-extension/issues/13269) is an important issue to track. Related issues [here](https://github.com/MetaMask/metamask-extension/issues/10623), [here](https://github.com/MetaMask/metamask-extension/issues/10597), and a related EIP discussion [here](https://ethereum-magicians.org/t/eip-3326-wallet-switchethereumchain/5471/13).

Here is a sample request that would fail the MetaMask check (pretending Alchemy is Casimir). 
```typescript
  async function addCasimirNode(provider: EthersProvider) {
    if (provider.request) {
      return await provider.request({
        method: 'wallet_addEthereumChain',
        params: [{
           chainId: '0x1',
           chainName: 'Ethereum (Casimir)',
           rpcUrls: [`https://eth-mainnet.alchemyapi.io/v2/${import.meta.env.PUBLIC_ALCHEMY_KEY}`],
           blockExplorerUrls: ['https://etherscan.io']
        }],
      })
    }
  }
```

We can, on the other hand, make the user complete the same steps manually, as MetaMask allows user-(in-wallet)-added custom RPC networks with default chains. See [here](https://docs.llama.fi/chainlist/how-to-change-ethereums-rpc).

**Comment by shanejearley** [comment]
Another example, this time using a non-default chain, so you can see how MetaMask looks when it works how we want it to work.

```typescript
  async function addBinanceNode(provider: EthersProvider) {
    if (provider.request) {
      return await provider.request({ 
        method: 'wallet_addEthereumChain', 
        params: [{ 
          chainId: '0x38', 
          chainName: 'Binance Smart Chain', 
          nativeCurrency: { name: 'BNB', symbol: 'BNB', decimals: 18 }, 
          rpcUrls: ['https://bsc-dataseed.binance.org/'], 
          blockExplorerUrls: ['https://bscscan.com/'] 
        }] 
      })
    }
  }
```

**Comment by ccali11** [comment]
### Ledger + Trezor Bitcoin App Address Types
Noting that it is worth considering which addresses users are expecting to use their ledger device with Casimir. Specifically, there are multiple types of addresses to consider: legacy (1), p2sh (segwit), bech32 (native segwit), etc. In my Ledger Live account, I only have one account and it is listed as "Native Segwit", but if I try to get the legacy address via Casimir, it will return me a legacy address. That could be confusing to the user, so we may want to give them the options to select which type of bitcoin address to choose.

**Reply by ccali11** [reply]
Read: 
[Ledger SegWit Support](https://www.ledger.com/ledger-releases-segregated-witness-support)
[Understanding difference btwn segwit and native segiwt](https://www.ledger.com/academy/difference-between-segwit-and-native-segwit)
[Ledger Crypto Assets Library](https://developers.ledger.com/docs/coin/cryptoassets-library/)

**Comment by ccali11** [comment]
### IoTeX Update

Chatted with IoTeX team in discord. They informed me the Desktop Wallet will soon be deprecated and to build for Metamask and Iopay Mobile. Thinking we integrate Iopay Mobile via WalletConnect given that it is a mobile wallet. When I asked them which staking methods to use with a web3 client library (e.g., ethers), they suggested taking a look at [this IIP](https://github.com/iotexproject/iips/blob/master/iip-12.md), but acknowledged they don't have any good examples of how to do this and that adding it to their documentation is in their backlog.

**Comment by ccali11** [comment]
### Coinbase Wallet Update

Noticing a few things to look into:
- [x] Wallet doesn't appear to be connected to Casimir (dapp) based on interface. Fix this.
- [ ] Wallet doesn't appear to support importing addresses (such as an IoTeX address). It may not even support this chain/token.
- [ ] Switching network to IoTeX network doesn't seem to throw an error, but chain does not switch.

**Reply by ccali11** [reply]
Revisited Coinbase Wallet today. There are some buggy things that I will log here:
- Coinbase Wallet still has MetaMask's old limitation of only returning one address when accessing like so: `const accounts = await provider.request({ method: 'eth_accounts' })`
- The active account is not easily determined; even when you've switched the active account in the Coinbase Wallet UI, the address that returns on request is not reliable (haven't figured out the logic that determines which one returns yet)
- There was some dapp connection and network bugginess that I'm not exactly sure how to replicate, but essentially I noticed Coinbase Wallet was a little unreliable when connecting and disconnecting the dapp (sometimes it would still allow me to interact with the dapp even after explicitly "disconnecting all dapp connections")
- I was able to sign from a non-active address even when using our `detectActiveWalletAddress" method

**Comment by ccali11** [comment]
### Ledger Wallet Update
1. Building bitcoin transactions is not all that well documented for [this library](https://github.com/LedgerHQ/ledger-live/tree/develop/libs/ledgerjs/packages/hw-app-btc) (hw-app-btc), so I asked around on the Ledger Discord and they said that library is old and pointed me to use [this](https://github.com/LedgerHQ/app-bitcoin-new/tree/develop/bitcoin_client_js) (app-bitcoin-new)
2. The `app-bitcoin-new` library was a little wonky (discussed with @shanejearley and he noticed the package is just not published or setup well for download and also very incomplete-looking and only ever had one commit back in March.
3. Used `npm install 'https://gitpkg.now.sh/LedgerHQ/app-bitcoin-new/bitcoin_client_js?develop'` to try and incorporate this in the project and use the methods they outline in that repo, but to no success. I keep getting the build error (see screenshot below), and tried the following three things:
   - include the dependency in vite config `optimizeDeps`
   - exclude the dependency in vite config `optimizeDeps`
   - cloning down the repository locally (got a 500 error, so not good)

We may need to just rely on old package from  LedgerHQ monorepo (from #1 above) for now to get address and balance, but no transactions for now. Will revisit for sure, but let me know if we want to prioritize @shanejearley

**Comment by shanejearley** [comment]
Sounds great Chris — good to move beyond it for the near term until we need BTC transactions. Thanks for noting.

**Comment by ccali11** [comment]
### Trezor Update
While it looks like Ledger allows you to select the Goerli Testnet App on the device itself before then connecting to the application, Trezor does not offer this sequence of steps. Instead, there is simply the Trezor Suite app that does not offer/require user to select which application they are choosing. Each time I try to connect using TrezorConnect library, it defaults to my Mainnet Ethereum account(s)/address(es), which is a different set of addresses than my Goerli Testnet account/address. My cursory research leads me to believe that we would have to programmatically alter the derivation path to account for selecting the Goerli Testnet because I noticed that Ropsten (and other testnets have different derivation paths), but then again Ledger does not seem to require different derivation paths to connect to the respective/intended applications (Mainnet vs. Testnet). This is an issue that is not critical for mainnet deployment, but would be critical for Goerli testnet deployment if we want to offer the user the ability to connect Trezor wallet.
